File: mobile/modules/frame-preview/ios/PreviewKit/Sources/FramePreviewCore/PreviewSlotPool.swift
import Foundation

/// One preallocated send buffer. Freed when the last reference goes, so a transport that still
/// holds a slot (through its send completion) keeps the memory alive after the pool replaced it.
public final class PreviewSlot {
  public let buffer: UnsafeMutableRawBufferPointer
  fileprivate let epoch: Int
  fileprivate var inUse = false
  fileprivate var sequence: UInt32?

  fileprivate init(byteCount: Int, epoch: Int) {
    buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: byteCount, alignment: 16)
    self.epoch = epoch
  }

  public var byteCount: Int { buffer.count }

  deinit { buffer.deallocate() }
}

/// Send buffers for packed frames: a fixed number of slots sized to the produced frame.
///
/// Slots are replaced only when the produced size changes — a tier change, or the source changing
/// resolution — and never per frame. A replaced slot still held by the worker or the transport is
/// retired rather than reused: its holder keeps the old memory, its late release is ignored, and
/// a new frame is never written over bytes still being read.
public final class PreviewSlotPool {
  private let slotCount: Int
  private var slots: [PreviewSlot] = []
  private var epoch = 0
  public private(set) var slotBytes = 0
  public private(set) var reallocations = 0

  public init(slotCount: Int = 2) {
    self.slotCount = max(slotCount, 1)
  }

  /// Allocate ahead of the first frame at a known size, off the pack path.
  public func prepare(byteCount: Int) {
    if byteCount > 0, byteCount != slotBytes { reshape(byteCount) }
  }

  /// A free slot of `byteCount` bytes, or nil when every slot is still held.
  public func acquire(byteCount: Int, sequence: UInt32) -> PreviewSlot? {
    if byteCount != slotBytes { reshape(byteCount) }
    for slot in slots where !slot.inUse {
      slot.inUse = true
      slot.sequence = sequence
      return slot
    }
    return nil
  }

  public func release(_ slot: PreviewSlot) {
    guard slot.epoch == epoch else { return }
    slot.inUse = false
    slot.sequence = nil
  }

  public func releaseAll() {
    for slot in slots {
      slot.inUse = false
      slot.sequence = nil
    }
  }

  public var heldCount: Int { slots.filter(\.inUse).count }

  private func reshape(_ byteCount: Int) {
    epoch += 1
    let current = epoch
    slots = (0 ..< slotCount).map { _ in PreviewSlot(byteCount: byteCount, epoch: current) }
    slotBytes = byteCount
    reallocations += 1
  }
}

File: mobile/modules/frame-preview/android/src/main/java/com/mentra/framepreview/PreviewSlotPool.kt
package com.mentra.framepreview

/**
 * Send buffers for packed frames: a fixed number of exactly-sized slots.
 *
 * Slots are sized to one produced frame because `JavaScriptReplyProxy.postMessage(byte[])` sends
 * the whole array. They are replaced only when the produced size changes — a tier change, or the
 * source changing resolution — and never per frame. A replaced slot that a worker or the
 * transport still holds is retired rather than reused: its holder keeps the old array and its
 * later release is ignored, so a new frame can never be written over bytes still being read.
 */
class PreviewSlotPool(private val slotCount: Int = 2) {
  class Slot internal constructor(val bytes: ByteArray, internal val epoch: Int) {
    internal var inUse = false
    internal var sequence = -1
  }

  private var slots: Array<Slot> = emptyArray()
  private var epoch = 0

  /** Bytes per slot in the current epoch, or 0 before the first allocation. */
  var slotBytes = 0
    private set

  /** Times the slot set was replaced. Stays flat while the produced size is steady. */
  var reallocations = 0
    private set

  /** Allocate ahead of the first frame at a known size, off the pack path. */
  @Synchronized
  fun prepare(byteCount: Int) {
    if (byteCount > 0 && byteCount != slotBytes) reshape(byteCount)
  }

  /** A free slot of exactly [byteCount] bytes, or null when every slot is still held. */
  @Synchronized
  fun acquire(byteCount: Int, sequence: Int): Slot? {
    if (byteCount != slotBytes) reshape(byteCount)
    for (slot in slots) {
      if (!slot.inUse) {
        slot.inUse = true
        slot.sequence = sequence
        return slot
      }
    }
    return null
  }

  @Synchronized
  fun release(slot: Slot) {
    if (slot.epoch != epoch) return
    slot.inUse = false
    slot.sequence = -1
  }

  /** Free the slot carrying [sequence], if it belongs to the current epoch. */
  @Synchronized
  fun releaseSequence(sequence: Int) {
    for (slot in slots) {
      if (slot.sequence == sequence) {
        slot.inUse = false
        slot.sequence = -1
      }
    }
  }

  @Synchronized
  fun releaseAll() {
    for (slot in slots) {
      slot.inUse = false
      slot.sequence = -1
    }
  }

  @Synchronized
  fun heldCount(): Int = slots.count { it.inUse }

  private fun reshape(byteCount: Int) {
    epoch += 1
    val current = epoch
    slots = Array(slotCount) { Slot(ByteArray(byteCount), current) }
    slotBytes = byteCount
    reallocations += 1
  }
}
